lib/ui/painting/immutable_buffer.cc

```diff
@@ -49,6 +53,50 @@ void ImmutableBuffer::init(Dart_NativeArguments args) {
   tonic::DartInvoke(callback_handle, {Dart_TypeVoid()});
 }
 
+void ImmutableBuffer::initFromAsset(Dart_NativeArguments args) {
+  UIDartState::ThrowIfUIOperationsProhibited();
+  Dart_Handle callback_handle = Dart_GetNativeArgument(args, 2);
+  if (!Dart_IsClosure(callback_handle)) {
+    Dart_SetReturnValue(args, tonic::ToDart("Callback must be a function"));
+    return;
+  }
+  Dart_Handle asset_name_handle = Dart_GetNativeArgument(args, 1);
+  uint8_t* chars = nullptr;
+  intptr_t asset_length = 0;
+  Dart_Handle result =
+      Dart_StringToUTF8(asset_name_handle, &chars, &asset_length);
+  if (Dart_IsError(result)) {
+    Dart_SetReturnValue(args, tonic::ToDart("Asset must be valid UTF8"));
+    return;
+  }
+  Dart_Handle immutable_buffer = Dart_GetNativeArgument(args, 0);
+
+  std::string asset_name = std::string{reinterpret_cast<const char*>(chars),
+                                       static_cast<size_t>(asset_length)};
+
+  std::shared_ptr<AssetManager> asset_manager = UIDartState::Current()
+                                                    ->platform_configuration()
+                                                    ->client()
+                                                    ->GetAssetManager();
+  std::unique_ptr<fml::Mapping> data = asset_manager->GetAsMapping(asset_name);
+  if (data == nullptr) {
+    Dart_SetReturnValue(args, tonic::ToDart("Asset not found"));
+    return;
+  }
+
+  const void* bytes = static_cast<const void*>(data->GetMapping());
+  auto size = data->GetSize();
+  void* peer = reinterpret_cast<void*>(data.release());
+  SkData::ReleaseProc proc = [](const void* ptr, void* context) {
+    auto* mapping = static_cast<fml::Mapping*>(const_cast<void*>(ptr));
```

Comment:
...something like this, will look next week

Reply:
so I think that the right thing to do would be to take a release proc from the mapping instance, and that could handle calling the right destructor. otherwise we don't know how to release the data